fix: remove randexp dependency

naresh-hopin opened the pull request:
* Removing the randexp dependency.
* randexp is removed from pact in this PR - https://github.com/pact-foundation/pact-ruby/pull/202 
* Issue with the randexp is that it overrides the Array's pick method in this file - https://github.com/benburkert/randexp/blob/master/lib/randexp/core_ext/array.rb 
* This is causing issues with other unit tests which we have in our repo.

Testing:
* I have tested pact-support after removing the randexp from requirement and looks fine from my end.

bethesque commented:
The randexp gem is used to generate a concrete value from a regular expression. It's unfortunate that there is no test to cover that, but it's used in this line https://github.com/pact-foundation/pact-support/blob/master/lib/pact/reification.rb#L17

If you can grab the relevant method from the randexp gem and copy it into the codebase, you can remove the dependency.

naresh-hopin commented:
Got it. Thanks, @bethesque :) 

I have replaced the randexp with [expgen](https://github.com/jnicklas/expgen). They have `There is a gem called Randexp which does much the same thing. Expgen differs from Randexp in two important ways. (1) It actually works. (2) It supports a much wider range of regexp syntax.` in their readme file which made me believe that it's an apt replacement. 

I have committed the code not sure how to trigger the unit tests to validate the change. Could you please help me with that?

bethesque commented:
I've actually found a test for it here: https://github.com/pact-foundation/pact-support/blob/master/spec/lib/pact/reification_spec.rb#L22

I don't know why it didn't fail when you removed the randexp dependency though 🤔

bethesque commented on lib/pact/reification.rb:
```diff
@@ -14,7 +14,7 @@ module Reification
       case term
       when Pact::Term, Regexp, Pact::SomethingLike, Pact::ArrayLike
-        from_term(term.generate)
+        from_term(Expgen.gen(term))
```
You'll need to split out the `Regexp` from the other classes, and only call `Expgen.gen(term)` for Regexp. The other classes still need to use `term.generate`.

naresh-hopin commented:
Hey @bethesque Thanks much for your patience and response on this. I have made a change to the gem which is used for `randexp` with `randomexp` where the monkey patch for the Array.pick is removed. I have run the unit tests locally and I could see those tests are passing. Can you please take a look again and if possible could we merge and push this as a new version, please. We are waiting on this badly for a long time. 

Thanks :) 

![Screenshot 2021-09-29 at 10 25 04 PM](https://user-images.githubusercontent.com/78914596/135314425-3f3fac4d-c761-4843-90ac-720907e05804.png)

bethesque commented:
Please revert this to what you had previously with the expgen gem, and just split out the `when Term` from the `when <other classes>` as I asked.

I'm very happy to help you get this fix out, but I won't be accepting a PR that uses a gem that you have copied, modified and then released without even publishing the [source code](https://github.com/nareshnavinash/randomexp). That's not a responsible decision as a maintainer.

naresh-hopin commented:
@bethesque Sorry for the confusion here. Changed as per your suggestion and updated the PR and I have validated the unit tests and they seem to be happy as well.

bethesque commented:
Perfect!

bethesque commented:
Gem is released - please update.

naresh-hopin commented:
Thanks much @bethesque

naresh-hopin commented:
@bethesque Can you also please take a look at this PR https://github.com/pact-foundation/pact-ruby/pull/247 where I have made a change in gemspec file of `pact-ruby` to update the version of pact-support. 🙇
